src/flint/types/fmpz_mod.pyx

```diff
@@ -0,0 +1,428 @@
+            val = any_as_fmpz(val)
+            if val is NotImplemented:
+                raise NotImplementedError("TODO")
+
+        fmpz_mod_set_fmpz(self.val, (<fmpz>val).val, self.ctx.val)
+
+    def is_zero(self):
+        return self == 0
+    
+    # TODO: kind of pointless, as we always ensure canonical on init?
+    def is_canonical(self):
+        cdef bint res
+        res = fmpz_mod_is_canonical(self.val, self.ctx.val)
+        return res == 1
+
+    def is_one(self):
+        cdef bint res
+        res = fmpz_mod_is_one(self.val, self.ctx.val)
+        return res == 1
+
+    def __richcmp__(s, t, int op):
+        cdef bint res
+        if op != 2 and op != 3:
+            raise TypeError("fmpz_mod cannot be ordered")
+        # TODO: is this the best method for comparison?
+        if typecheck(s, fmpz_mod) and typecheck(t, fmpz_mod):
+            res = ((<fmpz_mod>s).val[0] == (<fmpz_mod>t).val[0]) and \
+                  ((<fmpz_mod>s).ctx.val.n == (<fmpz_mod>t).ctx.val.n)
+            if op == 2:
+                return res
+            else:
+                return not res
+        # TODO: is this the best method for comparison?
+        # Seems like I'm doing too many type conversions?
+        elif typecheck(s, fmpz_mod) and typecheck(t, int):
+            res = int(s) == t % (<fmpz_mod>s).ctx.modulus()
+            if op == 2:
+                return res
+            else:
+                return not res
+        return NotImplemented
+
+    def __bool__(self):
+        return not self.is_zero()
+
+    def __repr__(self):
+        return "fmpz_mod({}, {})".format(
+            fmpz_get_intlong(self.val),
+            self.ctx.modulus()
+        )
+
+    # TODO: seems ugly...
+    def __hash__(self):
+        return  hash(
+            (int(self), int(self.ctx.modulus()))
+        )
+
+    def __int__(self):
+        return fmpz_get_intlong(self.val)
+
+    def __str__(self):
+        return str(int(self))
+
+    # ---------------- #
+    #    Arithmetic    #
+    # ---------------- #
+
+    def __pos__(self):
+        return self
+
+    def __neg__(self):
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
+
+        fmpz_mod_neg(
+            res.val, self.val, 
+            (<fmpz_mod_ctx_t>self.ctx.val))
+        return res
+
+    def __add__(self, other):
+        cdef fmpz_mod res
+        res = fmpz_mod.__new__(fmpz_mod)
+        res.ctx = self.ctx
+
+        # Add two fmpz_mod if moduli match
+        if typecheck(other, fmpz_mod):
+            if not self.ctx.val.n == (<fmpz_mod>other).ctx.val.n:
+                raise ValueError("moduli must match")
+
+            fmpz_mod_add(
+                res.val, self.val, (<fmpz_mod>other).val, self.ctx.val
+            )
+            return res
+
+        # Attempt to add an fmpz to an fmpz_mod
+        other = any_as_fmpz(other)
+        if other is NotImplemented:
+            raise NotImplementedError
```

Comment:
Methods like `__add__` should return `NotImplemented` for unrecognised types rather than raising `NotImplementedError`. The interpreter will translated `NotImplemented` into `TypeError` but only after both operands are tried i.e. if `a.__add__(b)` returns `NotImplemented` and `b.__radd__(a)` returns `NotImplemented` then `a + b` will give a `TypeError`. Raising `NotImplementedError` here means that `b.__radd__(a)` is not called and so not other type can overload `a + b` with `fmpz_mod`.